src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -91,13 +103,128 @@ private void Initialize(Compilation comp, AnalyzerExecutor analyzerExecutor, Can
+            }
+            finally
+            {
+                if (_initializeTask == null)
+                {
+                    // Set initializeTask to be a cancelled task.
+                    var tcs = new TaskCompletionSource<int>();
+                    tcs.SetCanceled();
+                    _initializeTask = tcs.Task;
+
+                    // Set primaryTask to be a cancelled task.
+                    tcs = new TaskCompletionSource<int>();
+                    tcs.SetCanceled();
+                    _primaryTask = tcs.Task;
+
+                    // Try to set the DiagnosticQueue to be complete.
+                    this.DiagnosticQueue.TryComplete();
+                }
+            }
+        }
+
+         internal void Initialize(
+            Compilation compilation,
+            AnalyzerOptions options,
+            Action<Exception, DiagnosticAnalyzer, Diagnostic> onAnalyzerException,
+            bool reportAnalyzer,
+            bool categorizeDiagnostics,
+            CancellationToken cancellationToken)
+        {
+            Debug.Assert(_initializeTask == null);
+
+            options = options ?? AnalyzerOptions.Empty;
+
+            var diagnosticQueue = DiagnosticQueue.Create(categorizeDiagnostics);
+            var addDiagnostic = GetDiagnosticSinkWithSuppression(diagnosticQueue.Enqueue, compilation);
+            var addLocalDiagnosticOpt = categorizeDiagnostics ? GetDiagnosticSinkWithSuppression(diagnosticQueue.EnqueueLocal, compilation) : null;
+            var addNonLocalDiagnosticOpt = categorizeDiagnostics ? GetDiagnosticSinkWithSuppression(diagnosticQueue.EnqueueNonLocal, compilation) : null;
+
+            Action<Exception, DiagnosticAnalyzer, Diagnostic> newOnAnalyzerException;
+            if (onAnalyzerException != null)
+            {
+                // Wrap onAnalyzerException to pass in filtered diagnostic.
+                var comp = compilation;
+                newOnAnalyzerException = (ex, analyzer, diagnostic) =>
+                    onAnalyzerException(ex, analyzer, GetFilteredDiagnostic(diagnostic, comp));
+            }
+            else
+            {
+                // Add exception diagnostic to regular diagnostic bag.
+                newOnAnalyzerException = (ex, analyzer, diagnostic) => addDiagnostic(diagnostic);
+            }
+
+            // Assume all analyzers are non-thread safe.
+            var singleThreadedAnalyzerToGateMap = ImmutableDictionary.CreateRange(_analyzers.Select(a => KeyValuePair.Create(a, new object())));
+
+            if (reportAnalyzer)
+            {
+                // If we are reporting detailed analyzer performance numbers, then do a dummy invocation of Compilation.GetTypeByMetadataName API upfront.
+                // This API seems to cause a severe hit for the first analyzer invoking it and hence introduces lot of noise in the computed analyzer execution times.
+                var unused = compilation.GetTypeByMetadataName("System.Object");
```

Comment:
use special type rather than System.Object. it can cause issue when there are multiple system.object.